staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -529,17 +529,10 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return nil, errors.NewServiceUnavailable(err.Error())
 	}
 
-	// determine the namespace and name scope of the watch, first from the request, secondarily from the field selector
-	scope := namespacedName{}
-	if requestNamespace, ok := request.NamespaceFrom(ctx); ok && len(requestNamespace) > 0 {
-		scope.namespace = requestNamespace
-	} else if selectorNamespace, ok := pred.Field.RequiresExactMatch("metadata.namespace"); ok {
-		scope.namespace = selectorNamespace
-	}
-	if requestInfo, ok := request.RequestInfoFrom(ctx); ok && requestInfo != nil && len(requestInfo.Name) > 0 {
+	var scope namespacedName
+	if requestInfo, ok := request.RequestInfoFrom(ctx); ok {
+		scope.namespace = requestInfo.Namespace
```

Comment:
I'm planning to add e2e tests for cases that Jordan laid out in his previous comment.

Reply:
I'm just not sure how to detect we hit the cache from an e2e test.

Reply:
> I think a unit test that can inspect the map of scopes → watchers might be easier than an e2e test here

+1